File: scala/15.scala
import scala.language.implicitConversions


implicit def int2bigint(n: Int) = BigInt(n)

def fact(n: BigInt): BigInt = (2 to n).reduceLeft(_ * _)

val num = fact(20 + 20)
val denom = fact(20) * fact(20)
val result = (num / denom).toLong

println(result)
assert(result == 137846528820L)

File: haskell/15.hs
-- time runghc 15.hs
-- 137846528820
-- 0.39user 0.03system 0:00.69elapsed 61%CPU (0avgtext+0avgdata 0maxresident)k
-- 2872inputs+0outputs (3major+6935minor)pagefaults 0swaps

fact n = product [1..n]
main = print result
       where num = fact 40
             denom = (fact 20) * (fact 20)
             result = div num denom
